src/main/java/patching/PatchManager.java

```diff
@@ -2,203 +2,433 @@
+                            logger.debug(String.format("PatchManager: Removing patches %s; patch queue is currently %s", batchingCtrl.patchBatchingQueue.subList(0, patches.length), batchingCtrl.patchBatchingQueue).replace("\n", "\\n"));
+                            logger.debug(String.format("PatchManager: Removing patches %s; patch done queue is currently %s", batchingCtrl.patchBatchingQueue.subList(0, patches.length), batchingCtrl.patchDoneQueue).replace("\n", "\\n"));
+
                             // Remove the sent patches
                             for (int i = 0; i < patches.length; i++) {
-                                batchingCtrl.patchDoneQueue.add(batchingCtrl.patchBatchingQueue.remove(0));
+                                batchingCtrl.patchBatchingQueue.remove(0);
                             }
 
+                            // Add server-acknowledged patches to doneQueue
+                            for (String change : ((FileChangeResponse) response.getData()).getChanges()) {
+                                batchingCtrl.patchDoneQueue.add(new Patch(change));
+                            }
+                            logger.debug(String.format("PatchManager: patch queue is currently %s, patch done queue is currently %s", batchingCtrl.patchBatchingQueue, batchingCtrl.patchDoneQueue).replace("\n", "\\n"));
                         }
+
+                        // Save missing patches & maxVersionSeen
                         if (((FileChangeResponse) response.getData()).getMissingPatches() != null) {
                             batchingCtrl.lastResponsePatches = ((FileChangeResponse) response.getData()).getMissingPatches();
                             batchingCtrl.maxVersionSeen = ((FileChangeResponse) response.getData()).getFileVersion();
                         }
+
+                        // Fire actual response handler
                         if (respHandler != null) {
                             respHandler.handleResponse(response);
                         }
                     }
 
                     logger.debug(String.format("PatchManager: File Change Success; running releaser. Changes sent: %s", Arrays.toString(patches)).replace("\n", "\\n") + "\n");
-                    releaser.run();
-                    timer.purge();
-                    timer.cancel();
+                    requestInFlightSem.release();
                 }, sendErrHandler
         );
 
         wsMgr.sendAuthenticatedRequest(req);
-        timer.schedule(new TimerTask() {
-            @Override
-            public void run() {
+
+        // Wait for up to PATCH_TIMEOUT_MILLIS for the response. After which, assume network failure and try again.
+        try {
+            long startTime = System.currentTimeMillis();
+            if (!requestInFlightSem.tryAcquire(PATCH_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS)) {
                 logger.debug("PatchManager: Request timed out, running releaser.");
```

Comment:
It still is; If the tryAcquire fails to get a permit within the timeout, it returns false. It then logs that debug statement